apps/prairielearn/src/pages/instructorInstanceAdminSettings/instructorInstanceAdminSettings.js

```diff
@@ -17,136 +15,125 @@ import {
-              );
-            },
-          );
-        }
-      });
-    });
-  } else if (req.body.__action === 'delete_course_instance') {
-    debug('Delete course instance');
-    const editor = new CourseInstanceDeleteEditor({
-      locals: res.locals,
-    });
-    editor.canEdit((err) => {
-      if (ERR(err, next)) return;
-      editor.doEdit((err, job_sequence_id) => {
-        if (ERR(err, (e) => logger.error('Error in doEdit()', e))) {
-          res.redirect(res.locals.urlPrefix + '/edit_error/' + job_sequence_id);
-        } else {
-          res.redirect(`${res.locals.plainUrlPrefix}/course/${res.locals.course.id}/course_admin`);
-        }
+    const host = getCanonicalHost(req);
+    res.locals.studentLink = new URL(
+      res.locals.plainUrlPrefix + '/course_instance/' + res.locals.course_instance.id,
+      host,
+    ).href;
+
+    res.locals.studentLinkQRCode = new QR({
+      content: res.locals.studentLink,
+      width: 512,
+      height: 512,
+    }).svg();
+
+    res.locals.infoCourseInstancePath = encodePath(
+      path.join(
+        'courseInstances',
+        res.locals.course_instance.short_name,
+        'infoCourseInstance.json',
+      ),
+    );
+    res.render(__filename.replace(/\.js$/, '.ejs'), res.locals);
+  }),
+);
+
+router.post(
+  '/',
+  asyncHandler(async (req, res) => {
+    if (req.body.__action === 'copy_course_instance') {
+      const editor = new CourseInstanceCopyEditor({
+        locals: res.locals,
       });
-    });
-  } else if (req.body.__action === 'change_id') {
-    debug(`Change short_name from ${res.locals.course_instance.short_name} to ${req.body.id}`);
-    if (!req.body.id) return next(new Error(`Invalid CIID (was falsy): ${req.body.id}`));
-    if (!/^[-A-Za-z0-9_/]+$/.test(req.body.id)) {
-      return next(
-        new Error(
-          `Invalid CIID (was not only letters, numbers, dashes, slashes, and underscores, with no spaces): ${req.body.id}`,
-        ),
+
+      const serverJob = await editor.prepareServerJob();
+      try {
+        await editor.executeWithServerJob(serverJob);
+      } catch (err) {
+        res.redirect(res.locals.urlPrefix + '/edit_error/' + serverJob.jobSequenceId);
+        return;
+      }
+
+      const courseInstanceID = await sqldb.queryRow(
+        sql.select_course_instance_id_from_uuid,
+        { uuid: editor.uuid, course_id: res.locals.course.id },
+        IdSchema,
       );
-    }
-    let ciid_new;
-    try {
-      ciid_new = path.normalize(req.body.id);
-    } catch (err) {
-      return next(new Error(`Invalid CIID (could not be normalized): ${req.body.id}`));
-    }
-    if (res.locals.course_instance.short_name === ciid_new) {
-      debug('The new ciid is the same as the old ciid - do nothing');
-      res.redirect(req.originalUrl);
-    } else {
-      const editor = new CourseInstanceRenameEditor({
+
+      if (courseInstanceID === null) {
+        throw new Error('Failed to find the new course instance ID');
+      }
+
```

Comment:
When would we ever hit this case? `queryRow` will throw an error if it doesn't find exactly one row.

Reply:
Good catch, removed in [6a9dc74](https://github.com/PrairieLearn/PrairieLearn/pull/9591/commits/6a9dc74dfafc5b4afe9791e93e16b59f1471c3dc).